workflow/controller/operator.go

```diff
@@ -246,13 +246,15 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 		}
 		woc.updated = wfUpdate
 		if !acquired {
-			woc.log.Warn("Workflow processing has been postponed due to concurrency limit")
-			phase := woc.wf.Status.Phase
-			if phase == wfv1.WorkflowUnknown {
-				phase = wfv1.WorkflowPending
+			if !woc.releaseLocksForPendingShuttingdownWfs(ctx) {
+				woc.log.Warn("Workflow processing has been postponed due to concurrency limit")
+				phase := woc.wf.Status.Phase
+				if phase == wfv1.WorkflowUnknown {
+					phase = wfv1.WorkflowPending
+				}
+				woc.markWorkflowPhase(ctx, phase, msg)
+				return
 			}
-			woc.markWorkflowPhase(ctx, phase, msg)
-			return
```

Comment:
this could be simplified / doesn't need another level of nesting

```suggestion
		if !acquired && !woc.releaseLocksForPendingShuttingdownWfs(ctx) {
			woc.log.Warn("Workflow processing has been postponed due to concurrency limit")
			phase := woc.wf.Status.Phase
			if phase == wfv1.WorkflowUnknown {
				phase = wfv1.WorkflowPending
			}
			woc.markWorkflowPhase(ctx, phase, msg)
			return
```

Simplified this in #12965